extensions/security/runtime/src/main/java/io/quarkus/security/runtime/interceptor/SecurityConstrainer.java

```diff
@@ -0,0 +1,164 @@
+            Class<?> declaringClass = method.getDeclaringClass();
+            securityAnnotation = determineSecurityAnnotation(declaringClass.getDeclaredAnnotations(),
+                    declaringClass::getCanonicalName);
+        }
+        if (securityAnnotation == null) {
+            securityAnnotation = determineSecurityAnnotationFromBindings(interceptorBindings, method::toString);
+        }
+        return checkForAnnotation(securityAnnotation);
+    }
+
+    private Optional<Check> checkForAnnotation(Annotation securityAnnotation) {
+        if (securityAnnotation instanceof DenyAll) {
+            return Optional.of(new DenyAllCheck());
+        }
+        if (securityAnnotation instanceof RolesAllowed) {
+            RolesAllowed rolesAllowed = (RolesAllowed) securityAnnotation;
+            return Optional.of(new RolesAllowedCheck(rolesAllowed.value()));
+        }
+        if (securityAnnotation instanceof PermitAll) {
+            return Optional.of(new PermitAllCheck());
+        }
+        if (securityAnnotation instanceof Authenticated) {
+            return Optional.of(new AuthenticatedCheck());
+        }
+        return Optional.empty();
+    }
+
+    private Annotation determineSecurityAnnotationFromBindings(Collection<Annotation> interceptorBindings,
+            Supplier<String> annotationPlacement) {
+        List<Annotation> securityAnnotations = interceptorBindings.stream()
+                .filter(anno -> SECURITY_ANNOTATIONS.stream().anyMatch(c -> c.isInstance(anno)))
+                .collect(Collectors.toList());
+        return getExactlyOne(securityAnnotations, annotationPlacement);
+    }
+
+    private Annotation determineSecurityAnnotation(Annotation[] annotations, Supplier<String> annotationPlacement) {
+        List<Annotation> securityAnnotations = Stream.of(annotations)
+                .filter(ann -> SECURITY_ANNOTATIONS.contains(ann.annotationType()))
+                .collect(Collectors.toList());
+
+        return getExactlyOne(securityAnnotations, annotationPlacement);
+    }
+
+    private Annotation getExactlyOne(List<Annotation> securityAnnotations, Supplier<String> annotationPlacement) {
+        switch (securityAnnotations.size()) {
+            case 0:
+                return null;
+            case 1:
+                return securityAnnotations.get(0);
+            default:
+                throw new IllegalStateException("Duplicate security annotations found on "
+                        + annotationPlacement.get() +
+                        ". Expected at most 1 annotation, found: " + securityAnnotations);
+        }
+    }
+
+    private static class RolesAllowedCheck implements Check {
+        private final String[] allowedRoles;
+
+        private RolesAllowedCheck(String[] allowedRoles) {
+            this.allowedRoles = allowedRoles;
+        }
+
+        @Override
+        public void apply(SecurityIdentity identity) {
+            if (identity.isAnonymous()) {
```

Comment:
I would skip this check. Potentially you can apply roles to unauthenticated identities (e.g. if a request comes from inside the company network you might want to give it slightly higher permissions). By default the role set will be empty anyway.

Reply:
don't we have to differentiate between 401 and 403?
Or should 401 be returned for a logged in user without roles too?

Reply:
Maybe do the check at the point you throw the exception.

Reply:
hmm... the exception is thrown here. Do you mean the point of translating the error to response?

Reply:
I mean change the code so it first checks the roles, as it is possible to have an anonymous identity that still has some roles (e.g. an anonymous identity from inside the company network might have more permissions than an external one). If the role check fails then decide which exception to throw.

Reply:
got it now, thanks.

Reply:
will be fixed in the next commit